alica_engine/src/engine/RunnableObject.cpp

```diff
@@ -34,70 +38,125 @@ ThreadSafePlanInterface RunnableObject::getPlanContext() const
 }
 
-void RunnableObject::initTrace()
+void RunnableObject::doInit()
 {
-    // Get closest parent that has a trace
-    if (areFlagsSet(Flags::TRACING_ENABLED) && _engine->getTraceFactory()) {
-        auto parent = _execContext.load()->getParent();
-        for (; parent && (!parent->getBasicPlan() || !parent->getBasicPlan()->getTraceContext().has_value()); parent = parent->getParent())
-            ;
-        _trace = _engine->getTraceFactory()->create(_name, parent ? parent->getBasicPlan()->getTraceContext() : std::nullopt);
+   ++_execState;
+
+    if (!isExecutingInContext()) {
+        return;
+    }
+    _initExecuted = true;
+
+    // There is a possible race condition here in the sense that the _execState can be behind the _signalState
+    // and yet this behaviour can execute in the _signalState's RunningPlan context. However this is harmless
+    // except for creating a superflous trace, since all other methods are guarded by isExecutingInContext()
+    // which will return false in all such cases.
+    // Atomically set the signal context to nullptr so that the RunningPlan instance can be deleted
+    // when the behaviour is terminated
+    _execContext = _signalContext.exchange(nullptr);
+
+    startTrace();
+
+    try {
+        if (_trace) {
+            _trace->setLog({"status", "initializing"});
+        }
+        onInit_();
+    } catch (const std::exception& e) {
+        ALICA_ERROR_MSG("[RunnableObject] Exception in: " << getName() << std::endl << e.what());
     }
 }
 
-void RunnableObject::traceRun()
+void RunnableObject::doRun()
 {
-    if (_trace && !areFlagsSet(Flags::RUN_TRACED)) {
-        _trace->setLog({"Run", "true"});
-        setFlags(Flags::RUN_TRACED);
+    // TODO: get rid of msg
+    try {
+        if (_trace && !_runTraced) {
+            _trace->setLog({"status", "running"});
+            _runTraced = true;
+        }
+        onRun_();
+    } catch (const std::exception& e) {
+        std::string err = std::string("Exception caught:  ") + getName() + std::string(" - ") + std::string(e.what());
+        sendLogMessage(4, err);
     }
 }
 
-void RunnableObject::traceInit(const std::string& type)
+void RunnableObject::doTerminate()
 {
-    if (_trace) {
-        _trace->setLog({type, "true"});
-        _trace->setLog({"Init", "true"});
+    ++_execState;
+
+    _runTraced = false;
+    if (!_initExecuted) {
+        // Reset the execution context so that the RunningPlan instance can be deleted
+        _execContext.store(nullptr);
+        return;
     }
+    _initExecuted = false;
+
+    // Intentionally call onTermination() at the end. This prevents setting success/failure from this method
+    try {
+        if (_trace) {
+            _trace->setLog({"status", "terminating"});
+        }
+        onTerminate_();
+        _trace.reset();
+    } catch (const std::exception& e) {
+        ALICA_ERROR_MSG("[BasicBehaviour] Exception in Behaviour-TERMINATE of: " << getName() << std::endl << e.what());
```

Comment:
nit: std::endl is kinda overused and better just to use `\n` unless you want to flush the output buffer at that point for some reason

Reply:
this is also a copy/paste of the old code. I'll remove the std::endl, do see any reason for using it here.